js/src/Ice/LoggerMiddleware.js

```diff
@@ -15,62 +16,115 @@ export class LoggerMiddleware {
+
+        DEV: console.assert(traceLevel > 0 || warningLevel > 0);
     }
 
-    warning(exceptionMessage, current) {
-        let output = "dispatch exception:";
-        output += `\nidentity: ${identityToString(current.id, this._toStringMode)}`;
-        output += `\nfacet: ${StringUtil.escapeString(current.facet, "", this._toStringMode)}`;
-        output += `\noperation: ${current.operation}`;
-        if (current.con !== null) {
-            try {
-                for (const p = current.con.getInfo(); p !== null; p = p.underlying) {
-                    if (p instanceof IPConnectionInfo) {
-                        output += `\nremote host: ${p.remoteAddress} remote port: ${p.remotePort}`;
-                        break;
-                    }
-                }
-            } catch (ex) {}
+    logDispatch(replyStatus, current) {
+        const output = [];
+        output.push("dispatch of ");
+        output.push(current.operation);
+        output.push(" to ");
+        printTarget(output, this._toStringMode, current);
+        output.push(" returned a response with reply status ");
+        output.push(replyStatus);
+        this._logger.trace(this._traceCat, output.join(""));
+    }
+
+    logDispatchFailed(exceptionDetails, current) {
+        const output = [];
+        output.push("failed to dispatch ");
+        output.push(current.operation);
+        output.push(" to ");
+        printTarget(output, this._toStringMode, current);
+
+        if (exceptionDetails !== null) {
+            output.push("\n");
+            output.push(exceptionDetails);
         }
+        this._logger.warning(output.join(""));
+    }
+}
+
+function printTarget(output, toStringMode, current) {
+    output.push("'");
+    output.push(identityToString(current.id, toStringMode));
+    output.push("'");
+    if (current.facet !== "") {
+        output.push(" -f '");
+        output.push(StringUtil.escapeString(current.facet, "", toStringMode));
+        output.push("'");
+    }
+    output.push(" over ");
+
+    DEV: console.assert(current.con !== null, "current.con is null"); // no colloc in JS
 
-        if (exceptionMessage !== null) {
-            output += `\n${exceptionMessage}`;
+    let connInfo = null;
+    try {
+        connInfo = current.con.getInfo();
+        while (connInfo.underlying !== null) {
+            connInfo = connInfo.underlying;
         }
-        this._logger.warning(output);
+    } catch (ex) {
+        // Thrown by getInfo() when the connection is closed.
+    }
+
+    if (connInfo instanceof IPConnectionInfo) {
+        output.push(connInfo.localAddress);
+        output.push(":");
+        output.push(connInfo.localPort);
+        output.push("<->");
+        output.push(connInfo.remoteAddress);
+        output.push(":");
+        output.push(connInfo.remotePort);
+    } else {
+        // Connection.toString() returns a multiline string, so we just use type here for bt and similar.
```

Comment:
I guess this would never happen in JavaScript we only have IP based transports.

Reply:
current.con.getInfo() can throw (because the connection is closed), resulting in a null connInfo. This could occur if the dispatch closes its own bidir connection.